OpenRA.Mods.D2k/Traits/SpiceBloom.cs

```diff
@@ -116,12 +118,10 @@ void ITick.Tick(Actor self)
 
 		void SeedResources(Actor self)
 		{
-			var pieces = self.World.SharedRandom.Next(info.Pieces[0], info.Pieces[1]) * ticks / growTicks;
-			if (pieces < info.Pieces[0])
-				pieces = info.Pieces[0];
-
-			var cells = self.World.Map.FindTilesInAnnulus(self.Location, 1, info.Range).ToList();
-
+			var pieces = int2.Lerp(info.Bursts[0], info.Bursts[1], ticks, growTicks);
+			var range = int2.Lerp(3, info.Range, ticks, growTicks);
+			var cells = self.World.Map.FindTilesInAnnulus(self.Location, 1, range).ToList();
+			var projectiles = new List<ProjectileArgs>();
```

Comment:
projectile creation should be in the FireSequence. Things like CanAddResource will change as time goes on

Reply:
What is the best solution here? Only solution I can think of is to pass all Spicebloom info into FireSequence and do all SeedResources there.

Reply:
That would probably be the cleanest way. Then you could rename the effect to FireSpiceBloomEffect

Reply:
After some digging I think this is no go zone. You will need pass whole Spicebloom instance with all Modifiers, Location etc. Which will die tick later and its the opposite to why FireSequence exists in first place. 
If `CanAddResource `is only concern it will be easier just recheck it before projectile is spawned.

Reply:
just lock in all values and only keep this dynamic

```csharp
				var cell = cells
					.SkipWhile(p => resourceLayer.GetResource(p).Type == info.ResourceType && !resourceLayer.CanAddResource(info.ResourceType, p))
					.Cast<CPos?>()
					.RandomOrDefault(self.World.SharedRandom);

				if (cell == null)
					cell = cells.Random(self.World.SharedRandom);
```

also this is implemented poorly, it's more effective as something like this

```csharp
				var emptyCells = cells
					.Where(p =>
						resourceLayer.GetResource(p).Type != info.ResourceType
						|| resourceLayer.CanAddResource(info.ResourceType, p))
					.ToList();

				var cell = emptyCells.Count == 0
					? cells.RandomOrDefault(self.World.SharedRandom)
					: emptyCells.RandomOrDefault(self.World.SharedRandom);
```

Reply:
Actually i'm not sure if this needs solving at all - why is the cell becoming invalid 1 tick before the projectile is fired important to handle if the cell becoming invalid 1 cell after the projectile is fired isn't?

Reply:
I know, I experiment with this a while. But at the end I decide it should be mapmaker responsibility to place spicebloom in area where it can spread. Trait responsibility should be that it should spread equally.

Reply:
then at least use a Stack to hold projectiles instead of a List